chalk-engine/src/simplify.rs

```diff
@@ -61,13 +62,42 @@ impl<I: Interner, C: Context<I>> Forest<I, C> {
                             subgoal.clone(),
                         )));
                 }
-                GoalData::EqGoal(goal) => infer.unify_generic_args_into_ex_clause(
+                GoalData::EqGoal(goal) => match infer.relate_generic_args_into_ex_clause(
                     context.interner(),
+                    context.unification_database(),
                     &environment,
+                    Variance::Invariant,
                     &goal.a,
                     &goal.b,
                     &mut ex_clause,
-                )?,
+                ) {
+                    Ok(()) => {}
+                    Err(_) => return FallibleOrFloundered::NoSolution,
+                },
+                GoalData::SubtypeGoal(goal) => {
+                    match (goal.a.ty(context.interner()), goal.b.ty(context.interner())) {
+                        (Some(a_ty), Some(b_ty)) => {
+                            if a_ty.inference_var(context.interner()).is_some()
+                                && b_ty.inference_var(context.interner()).is_some()
+                            {
+                                return FallibleOrFloundered::Floundered;
+                            }
+                        }
+                        _ => {}
+                    }
+                    match infer.relate_generic_args_into_ex_clause(
```

Comment:
So, I made `SubtypeGoal` actually have two `GenericArg`s to be able to reuse this function. (And similarly the one in `recursive`). But I think I just might change this function to be generic over `Zip`